crates/miden-objects/src/account/component/metadata/storage/schema.rs

```diff
@@ -0,0 +1,1202 @@
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+use core::iter;
+
+use miden_core::utils::{ByteReader, ByteWriter, Deserializable, Serializable};
+use miden_processor::DeserializationError;
+
+use super::type_registry::{InitValueRequirement, SCHEMA_TYPE_REGISTRY, SchemaTypeIdentifier};
+use super::{InitStorageData, StorageValueName, WordValue};
+use crate::account::{AccountStorage, StorageMap, StorageSlot, StorageSlotName};
+use crate::errors::AccountComponentTemplateError;
+use crate::{Felt, FieldElement, Word};
+
+/// Alias used for iterators that collect all init-required values and their types within a
+/// component storage schema.
+pub type SchemaRequirementsIter<'a> =
+    Box<dyn Iterator<Item = (StorageValueName, InitValueRequirement)> + 'a>;
+
+// STORAGE SCHEMA
+// ================================================================================================
+
+/// Describes the storage layout of an account component in terms of named storage slots.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct AccountStorageSchema {
+    fields: BTreeMap<StorageSlotName, StorageSlotSchema>,
+}
+
+impl AccountStorageSchema {
+    /// Creates a new [`AccountStorageSchema`].
+    ///
+    /// # Errors
+    /// - If `fields` contains duplicate slot names.
+    /// - If `fields` contains the protocol-reserved faucet metadata slot name.
+    pub fn new(
+        fields: impl IntoIterator<Item = (StorageSlotName, StorageSlotSchema)>,
+    ) -> Result<Self, AccountComponentTemplateError> {
+        let mut map = BTreeMap::new();
+        for (slot_name, schema) in fields {
+            if slot_name.id() == AccountStorage::faucet_sysdata_slot().id() {
+                return Err(AccountComponentTemplateError::ReservedSlotName(slot_name));
+            }
+
+            if map.insert(slot_name.clone(), schema).is_some() {
+                return Err(AccountComponentTemplateError::DuplicateSlotName(slot_name));
+            }
+        }
+
+        Ok(Self { fields: map })
+    }
+
+    /// Returns an iterator over `(slot_name, schema)` pairs in slot-id order.
+    pub fn iter(&self) -> impl Iterator<Item = (&StorageSlotName, &StorageSlotSchema)> {
+        self.fields.iter()
+    }
+
+    /// Returns a reference to the underlying fields map.
+    pub fn fields(&self) -> &BTreeMap<StorageSlotName, StorageSlotSchema> {
+        &self.fields
+    }
+
+    /// Builds the initial [`StorageSlot`]s for this schema using the provided initialization data.
+    pub fn build_storage_slots(
+        &self,
+        init_storage_data: &InitStorageData,
+    ) -> Result<Vec<StorageSlot>, AccountComponentTemplateError> {
+        self.fields
+            .iter()
+            .map(|(slot_name, schema)| schema.try_build_storage_slot(slot_name, init_storage_data))
+            .collect()
+    }
+
+    /// Returns an iterator over init-value requirements for the entire schema.
+    pub fn init_value_requirements(&self) -> SchemaRequirementsIter<'_> {
```

Comment:
Nit: Sometimes these methods seem to be called `schema_requirements` and sometimes `init_value_requirements`, should that be consistent? If so, I would prefer the former.